File: PE/AdatStrucc/2025_02_26_Recursion/partition.ml
let rec partition_max number max_number =
    match (number, max_number) with
    | (_, 1) -> 1
    | (1, _) -> 1
    | (n, k) when k>n -> partition_max n n
    | (n, k) when n==k -> 1 + partition_max n (n - 1)
    | (n, k) -> partition_max n (k - 1) + partition_max (n - k) k
;;

print_int (partition_max 10 10);
print_int (partition_max 100 100);
print_newline ();;

File: PE/AdatStrucc/2025_02_26_Recursion/partition.cpp
#include <iostream>

int partition_max(int n, int k) {
    if (k==1) return 1;
    if (n==1) return 1;
    if (k>n) return partition_max(n,n);
    if (k==n) return 1 + partition_max(n,n-1);
    return partition_max(n-k,k) + partition_max(n,k-1);
}

int main() {
    std::cout << partition_max(10,10) << std::endl;
    std::cout << partition_max(100,100) << std::endl;
}